### ISSUE TITLE
Update S2971: Rule should not only suggest to remove call to ToList or ToArray

### ISSUE THREAD

**Issue by papaneko** [SonarSource/sonar-csharp#1711]
### Description

Using ToList() in the middle of  a LINQ to Entities call chain query should be replaced with AsEnumerable() 
instead of just being dropped.

### Repro steps
1. Step A
Run analysis on a LINQ to Entities query

```
context.MyEntities.Where(x => x.Foo == 'bar').ToList().Select(e => new Dto 
{ 
   MyProperty = GetValue(e) 
})} 

public int GetValue(Foo bar){ ... }
```

2. Step B

### Expected behavior

Message from S2971 should be something similiar to "Replace ToList() with AsEnumerable()"

Since materializing the query to memory is necessary to do before running GetValue

### Actual behavior

Sonarqube reports 
"Drop 'ToList' from the middle of the call chain. (S2971)"

Dropping ToList from the call chain will cause a NotSupportedException (LINQ to Entities)

### Known workarounds

Please provide a description of any known workarounds.

### Related information

* SonarC# Version
* Visual Studio Version
* If running through the Scanner for MSBuild, its version

**Comment by Evangelink** [comment]
Hi @papaneko,

I haven't played with databases query in quite some time but I recall that I was doing a `AsEnumerable<T>` call instead of this `ToList`. And I now we don't report for the `AsEnumerable`.

Is there any reason to force the `ToList` instead of the `AsEnumerable`?

**Comment by papaneko** [comment]
Ahh, yes you are correct. I should have been using AsEnumerable() instead of ToList(), since i don't want to create an unnecessary intermedate list.

Then i would suggest another expected behaviour:
"Replace ToList() with AsEnumerable()" 

Since dropping ToList() from the query will make it throw a NotSupportedException (LINQ to Entities)

I have not tried running the query with AsEnumerable() through SonarQube.

**Comment by Evangelink** [comment]
You are right, it's better if the rule suggest something like:
`Either replace this call to 'ToList' by 'AsEnumerable' when using LINQ to Entities or drop it for LINQ to object`.

WDYT?

**Comment by papaneko** [comment]
Yes! Sounds good.
